Cython/Compiler/ExprNodes.py

```diff
@@ -2917,6 +2919,20 @@ def analyse_types(self, env):
                 IdentifierStringNode(self.module_name.pos, value=part)
                 for part in map(StringEncoding.EncodedString, self.module_name.value.split('.'))
             ]).analyse_types(env)
+        if self.level > 0:
+            qualname = env.global_scope().qualified_name
+            for _ in range(self.level):
+                qualname, _, _ = qualname.rpartition('.')
+            qualname = IdentifierStringNode(
+                self.module_name.pos,
+                value=StringEncoding.EncodedString(f"{qualname}.{self.module_name.value}"),
+            ).analyse_types(env)
+        else:
+            qualname = IdentifierStringNode(
+                self.module_name.pos,
+                value=StringEncoding.EncodedString(self.module_name.value),
+            ).analyse_types(env)
```

Comment:
I don't know if this makes sense for the `level=-1` case (which admittedly is trying to emulated old behaviour).
Again I think the solution is probably not to pass a qualname and to skip the optimized lookup.